File: Sources/Ably/ARTConnection.m
//
//  ARTConnection.m
//  ably
//
//  Created by Ricardo Pereira on 30/10/2015.
//  Copyright © 2015 Ably. All rights reserved.
//

#import "ARTConnection+Private.h"

#import "ARTRealtime+Private.h"
#import "ARTEventEmitter+Private.h"
#import "ARTQueuedDealloc.h"

@implementation ARTConnection {
    ARTQueuedDealloc *_dealloc;
}

- (ARTConnectionInternal *_Nonnull)internal_nosync {
    return _internal;
}

- (NSString *)id {
    return _internal.id;
}

- (NSString *)key {
    return _internal.key;
}

- (NSString *)recoveryKey {
    return _internal.recoveryKey;
}

- (int64_t)serial {
    return _internal.serial;
}

- (NSInteger)maxMessageSize {
    return _internal.maxMessageSize;
}

- (ARTRealtimeConnectionState)state {
    return _internal.state;
}

- (ARTErrorInfo *)errorReason {
    return _internal.errorReason;
}

- (instancetype)initWithInternal:(ARTConnectionInternal *)internal queuedDealloc:(ARTQueuedDealloc *)dealloc {
    self = [super init];
    if (self) {
        _internal = internal;
        _dealloc = dealloc;
    }
    return self;
}

- (void)close {
    [_internal close];
}

- (void)connect {
    [_internal connect];
}

- (void)off {
    [_internal off];
}

- (void)off:(nonnull ARTEventListener *)listener {
    [_internal off:listener];
}

- (void)off:(ARTRealtimeConnectionEvent)event listener:(nonnull ARTEventListener *)listener {
    [_internal off:event listener:listener];
}

- (nonnull ARTEventListener *)on:(nonnull void (^)(ARTConnectionStateChange * _Nullable))cb {
    return [_internal on:cb];
}

- (nonnull ARTEventListener *)on:(ARTRealtimeConnectionEvent)event callback:(nonnull void (^)(ARTConnectionStateChange * _Nullable))cb {
    return [_internal on:event callback:cb];
}

- (nonnull ARTEventListener *)once:(nonnull void (^)(ARTConnectionStateChange * _Nullable))cb {
    return [_internal once:cb];
}

- (nonnull ARTEventListener *)once:(ARTRealtimeConnectionEvent)event callback:(nonnull void (^)(ARTConnectionStateChange * _Nullable))cb {
    return [_internal once:event callback:cb];
}

- (void)ping:(nonnull void (^)(ARTErrorInfo * _Nullable))cb {
    [_internal ping:cb];
}

@end

@implementation ARTConnectionInternal {
    _Nonnull dispatch_queue_t _queue;
    NSString *_id;
    NSString *_key;
    NSInteger _maxMessageSize;
    int64_t _serial;
    ARTRealtimeConnectionState _state;
    ARTErrorInfo *_errorReason;
}

- (instancetype)initWithRealtime:(ARTRealtimeInternal *)realtime {
    if (self = [super init]) {
        _eventEmitter = [[ARTPublicEventEmitter alloc] initWithRest:realtime.rest];
        _realtime = realtime;
        _queue = _realtime.rest.queue;
        _serial = -1;
    }
    return self;
}

- (void)connect {
    [_realtime connect];
}

- (void)close {
    [_realtime close];
}

- (void)ping:(void (^)(ARTErrorInfo *))cb {
    [_realtime ping:cb];
}

- (NSString *)id {
    __block NSString *ret;   
dispatch_sync(_queue, ^{
    ret = [self id_nosync];
});
    return ret;
} 

- (NSString *)key {
    __block NSString *ret;   
dispatch_sync(_queue, ^{
    ret = [self key_nosync];
});
    return ret;
} 

- (int64_t)serial {
    __block int64_t ret;   
dispatch_sync(_queue, ^{
    ret = [self serial_nosync];
});
    return ret;
} 

- (ARTRealtimeConnectionState)state {
    __block ARTRealtimeConnectionState ret;   
dispatch_sync(_queue, ^{
    ret = [self state_nosync];
});
    return ret;
}

- (ARTErrorInfo *)errorReason {
    __block ARTErrorInfo *ret;   
dispatch_sync(_queue, ^{
    ret = [self errorReason_nosync];
});
    return ret;
}

- (ARTErrorInfo *)error_nosync {
    if (self.errorReason_nosync) {
        return self.errorReason_nosync;
    }
    switch (self.state_nosync) {
        case ARTRealtimeDisconnected:
            return [ARTErrorInfo createWithCode:80003 status:400 message:@"Connection to server temporarily unavailable"];
        case ARTRealtimeSuspended:
            return [ARTErrorInfo createWithCode:80002 status:400 message:@"Connection to server unavailable"];
        case ARTRealtimeFailed:
            return [ARTErrorInfo createWithCode:80000 status:400 message:@"Connection failed or disconnected by server"];
        case ARTRealtimeClosing:
            return [ARTErrorInfo createWithCode:80017 status:400 message:@"Connection closing"];
        case ARTRealtimeClosed:
            return [ARTErrorInfo createWithCode:80003 status:400 message:@"Connection closed"];
        default:
            return [ARTErrorInfo createWithCode:80010 status:400 message:[NSString stringWithFormat:@"Invalid operation (connection state is %lu - %@)", (unsigned long)self.state_nosync, ARTRealtimeConnectionStateToStr(self.state_nosync)]];
    }
}

- (BOOL)isActive_nosync {
    return self.realtime.isActive;
}

- (NSString *)id_nosync {
    return _id;
} 

- (NSString *)key_nosync {
    return _key;
} 

- (int64_t)serial_nosync {
    return _serial;
} 

- (ARTRealtimeConnectionState)state_nosync {
    return _state;
}

- (ARTErrorInfo *)errorReason_nosync {
    return _errorReason;
}

- (void)setId:(NSString *)newId {
    _id = newId;
}

- (void)setKey:(NSString *)key {
    _key = key;
}

- (void)setSerial:(int64_t)serial {
    _serial = serial;
}

- (void)setMaxMessageSize:(NSInteger)maxMessageSize {
    _maxMessageSize = maxMessageSize;
}

- (void)setState:(ARTRealtimeConnectionState)state {
    _state = state;
}

- (void)setErrorReason:(ARTErrorInfo *_Nullable)errorReason {
    _errorReason = errorReason;
}

- (NSString *)recoveryKey {
    __block NSString *ret;
dispatch_sync(_queue, ^{
    ret = [self recoveryKey_nosync];
});
    return ret;
}

- (NSString *)recoveryKey_nosync {
    switch(self.state_nosync) {
        case ARTRealtimeConnecting:
        case ARTRealtimeConnected:
        case ARTRealtimeDisconnected:
        case ARTRealtimeSuspended: {
            NSString *recStr = self.key_nosync;
            if (recStr == nil) {
                return nil;
            }
            NSString *str = [recStr stringByAppendingString:[NSString stringWithFormat:@":%ld:%ld", (long)self.serial_nosync, (long)_realtime.msgSerial]];
            return str;
        } default:
            return nil;
    }
}

- (ARTEventListener *)on:(ARTRealtimeConnectionEvent)event callback:(void (^)(ARTConnectionStateChange *))cb {
    return [_eventEmitter on:[ARTEvent newWithConnectionEvent:event] callback:cb];
}

- (ARTEventListener *)on:(void (^)(ARTConnectionStateChange *))cb {
    return [_eventEmitter on:cb];
}

- (ARTEventListener *)once:(ARTRealtimeConnectionEvent)event callback:(void (^)(ARTConnectionStateChange *))cb {
    return [_eventEmitter once:[ARTEvent newWithConnectionEvent:event] callback:cb];
}

- (ARTEventListener *)once:(void (^)(ARTConnectionStateChange *))cb {
    return [_eventEmitter once:cb];
}

- (void)off {
    if (_realtime && _realtime.rest) {
        [_eventEmitter off];
    } else {
        [_eventEmitter off];
    }
}
- (void)off:(ARTRealtimeConnectionEvent)event listener:(ARTEventListener *)listener {
    [_eventEmitter off:[ARTEvent newWithConnectionEvent:event] listener:listener];
}

- (void)off:(ARTEventListener *)listener {
    [_eventEmitter off:listener];
}

- (void)emit:(ARTRealtimeConnectionEvent)event with:(ARTConnectionStateChange *)data {
    [_eventEmitter emit:[ARTEvent newWithConnectionEvent:event] with:data];
}

@end

#pragma mark - ARTEvent

@implementation ARTEvent (ConnectionEvent)

- (instancetype)initWithConnectionEvent:(ARTRealtimeConnectionEvent)value {
    return [self initWithString:[NSString stringWithFormat:@"ARTRealtimeConnectionEvent%@", ARTRealtimeConnectionEventToStr(value)]];
}

+ (instancetype)newWithConnectionEvent:(ARTRealtimeConnectionEvent)value {
    return [[self alloc] initWithConnectionEvent:value];
}

@end

File: Sources/AblySwift/Transport/ARTConnection.swift
//
//  ARTConnection.swift
//  AblySwift
//
//  Created during Swift migration from Objective-C.
//  Copyright © 2024 Ably Real-time Ltd. All rights reserved.
//

import Foundation

// MARK: - Connection Recovery Key

public class ARTConnectionRecoveryKey: NSObject, @unchecked Sendable {
    
    public let connectionKey: String
    public let msgSerial: Int64
    public let channelSerials: [String: String]
    
    public init(connectionKey: String, msgSerial: Int64, channelSerials: [String: String]) {
        self.connectionKey = connectionKey
        self.msgSerial = msgSerial
        self.channelSerials = channelSerials
        super.init()
    }
    
    public func jsonString() -> String {
        let object: [String: Any] = [
            "msgSerial": msgSerial,
            "connectionKey": connectionKey,
            "channelSerials": channelSerials
        ]
        
        do {
            let jsonData = try JSONSerialization.data(withJSONObject: object, options: [])
            return String(data: jsonData, encoding: .utf8) ?? ""
        } catch {
            fatalError("ARTConnectionRecoveryKey: This JSON serialization should pass without errors.")
        }
    }
    
    public static func fromJsonString(_ json: String) -> (ARTConnectionRecoveryKey?, Error?) {
        guard let jsonData = json.data(using: .utf8) else {
            return (nil, NSError(domain: "ARTConnectionRecoveryKey", code: -1, userInfo: [NSLocalizedDescriptionKey: "Invalid JSON string"]))
        }
        
        do {
            guard let object = try JSONSerialization.jsonObject(with: jsonData, options: []) as? [String: Any] else {
                return (nil, NSError(domain: "ARTConnectionRecoveryKey", code: -1, userInfo: [NSLocalizedDescriptionKey: "Invalid JSON format"]))
            }
            
            let connectionKey = object["connectionKey"] as? String ?? ""
            let msgSerial = object["msgSerial"] as? NSNumber ?? 0
            let channelSerials = object["channelSerials"] as? [String: String] ?? [:]
            
            let recoveryKey = ARTConnectionRecoveryKey(
                connectionKey: connectionKey,
                msgSerial: msgSerial.int64Value,
                channelSerials: channelSerials
            )
            
            return (recoveryKey, nil)
        } catch {
            return (nil, error)
        }
    }
}

// MARK: - Connection Internal

internal class ARTConnectionInternal: NSObject, @unchecked Sendable {
    
    // MARK: - Properties
    
    private let queue: DispatchQueue
    private var _id: String?
    private var _key: String?
    private var _maxMessageSize: Int = 0
    private var _state: ARTRealtimeConnectionState = .initialized
    private var _errorReason: ARTErrorInfo?
    
    internal let eventEmitter: ARTEventEmitter<ARTEvent, ARTConnectionStateChange>
    internal weak var realtime: ARTRealtimeInternal?
    
    // MARK: - Initializers
    
    internal init(realtime: ARTRealtimeInternal, logger: ARTInternalLog) {
        self.queue = DispatchQueue(label: "com.ably.connection", qos: .default)
        self.eventEmitter = ARTEventEmitter<ARTEvent, ARTConnectionStateChange>()
        self.realtime = realtime
        super.init()
    }
    
    // MARK: - Public Properties (Thread-Safe)
    
    internal var id: String? {
        return queue.sync { _id }
    }
    
    internal var key: String? {
        return queue.sync { _key }
    }
    
    internal var maxMessageSize: Int {
        return queue.sync {
            if _maxMessageSize > 0 {
                return _maxMessageSize
            }
            // Return default based on environment - placeholder logic
            return 65536 // Default max message size
        }
    }
    
    internal var state: ARTRealtimeConnectionState {
        return queue.sync { _state }
    }
    
    internal var errorReason: ARTErrorInfo? {
        return queue.sync { _errorReason }
    }
    
    // MARK: - Nosync Properties (Queue Already Held)
    
    internal var id_nosync: String? { return _id }
    internal var key_nosync: String? { return _key }
    internal var state_nosync: ARTRealtimeConnectionState { return _state }
    internal var errorReason_nosync: ARTErrorInfo? { return _errorReason }
    
    internal var isActive_nosync: Bool {
        return realtime?.isActive ?? false
    }
    
    internal var error_nosync: ARTErrorInfo? {
        if let errorReason = _errorReason {
            return errorReason
        }
        
        switch _state {
        case .disconnected:
            return ARTErrorInfo.create(withCode: 80003, message: "Connection to server temporarily unavailable")
        case .suspended:
            return ARTErrorInfo.create(withCode: 80002, message: "Connection to server unavailable")
        case .failed:
            return ARTErrorInfo.create(withCode: 80000, message: "Connection failed or disconnected by server")
        case .closing:
            return ARTErrorInfo.create(withCode: 80017, message: "Connection closing")
        case .closed:
            return ARTErrorInfo.create(withCode: 80003, message: "Connection closed")
        default:
            return ARTErrorInfo.create(withCode: 80013, message: "Invalid operation (connection state is \(_state.rawValue) - \(ARTRealtimeConnectionStateToStr(_state)))")
        }
    }
    
    // MARK: - Setters (Thread-Safe)
    
    internal func setId(_ newId: String?) {
        queue.sync { _id = newId }
    }
    
    internal func setKey(_ key: String?) {
        queue.sync { _key = key }
    }
    
    internal func setMaxMessageSize(_ maxMessageSize: Int) {
        queue.sync { _maxMessageSize = maxMessageSize }
    }
    
    internal func setState(_ state: ARTRealtimeConnectionState) {
        queue.sync {
            _state = state
            if isInactiveConnectionState(state) {
                _id = nil
                _key = nil
            }
        }
    }
    
    internal func setErrorReason(_ errorReason: ARTErrorInfo?) {
        queue.sync { _errorReason = errorReason }
    }
    
    // MARK: - Recovery Key
    
    @available(*, deprecated, message: "Use createRecoveryKey method instead.")
    internal var recoveryKey: String? {
        return createRecoveryKey()
    }
    
    internal func createRecoveryKey() -> String? {
        return queue.sync { createRecoveryKey_nosync() }
    }
    
    internal func createRecoveryKey_nosync() -> String? {
        guard let key = _key, !isInactiveConnectionState(_state) else {
            return nil
        }
        
        var channelSerials: [String: String] = [:]
        
        // Placeholder for channel serials collection
        // Will be implemented when realtime channels are available
        
        let recoveryKey = ARTConnectionRecoveryKey(
            connectionKey: key,
            msgSerial: 0, // Placeholder - will use realtime.msgSerial
            channelSerials: channelSerials
        )
        
        return recoveryKey.jsonString()
    }
    
    // MARK: - Connection Operations
    
    internal func connect() {
        realtime?.connect()
    }
    
    internal func close() {
        realtime?.close()
    }
    
    internal func ping(_ callback: @escaping ARTCallback) {
        realtime?.ping(callback)
    }
    
    // MARK: - Event Emitter Methods
    
    internal func emit(_ event: ARTRealtimeConnectionEvent, with data: ARTConnectionStateChange) {
        let artEvent = ARTEvent(connectionEvent: event)
        eventEmitter.emit(artEvent, with: data)
    }
    
    // MARK: - Helper Methods
    
    private func isInactiveConnectionState(_ state: ARTRealtimeConnectionState) -> Bool {
        return state == .closing || state == .closed || state == .failed || state == .suspended
    }
}

// MARK: - Public Connection

public class ARTConnection: NSObject, @unchecked Sendable {
    
    internal let _internal: ARTConnectionInternal
    private let _dealloc: ARTQueuedDealloc?
    
    // MARK: - Initializers
    
    internal init(internal: ARTConnectionInternal, queuedDealloc: ARTQueuedDealloc?) {
        self._internal = `internal`
        self._dealloc = queuedDealloc
        super.init()
    }
    
    // MARK: - Public Properties
    
    public var id: String? {
        return _internal.id
    }
    
    public var key: String? {
        return _internal.key
    }
    
    public var maxMessageSize: Int {
        return _internal.maxMessageSize
    }
    
    public var state: ARTRealtimeConnectionState {
        return _internal.state
    }
    
    public var errorReason: ARTErrorInfo? {
        return _internal.errorReason
    }
    
    @available(*, deprecated, message: "Use createRecoveryKey method instead.")
    public var recoveryKey: String? {
        return _internal.recoveryKey
    }
    
    // MARK: - Public Methods
    
    public func createRecoveryKey() -> String? {
        return _internal.createRecoveryKey()
    }
    
    public func connect() {
        _internal.connect()
    }
    
    public func close() {
        _internal.close()
    }
    
    public func ping(_ callback: @escaping ARTCallback) {
        _internal.ping(callback)
    }
    
    // MARK: - Internal Access
    
    internal var internal_nosync: ARTConnectionInternal {
        return _internal
    }
}

// MARK: - Event Extensions

extension ARTEvent {
    convenience init(connectionEvent: ARTRealtimeConnectionEvent) {
        self.init()
        // Store the connection event - placeholder implementation for now
    }
    
    static func newWithConnectionEvent(_ event: ARTRealtimeConnectionEvent) -> ARTEvent {
        return ARTEvent(connectionEvent: event)
    }
}

// MARK: - Forward Declarations

// Placeholder for ARTRealtimeInternal - will be migrated in Phase 7
internal class ARTRealtimeInternal: @unchecked Sendable {
    internal var isActive: Bool { return false }
    internal func connect() {}
    internal func close() {}
    internal func ping(_ callback: @escaping ARTCallback) {
        callback(nil)
    }
}